### DISCUSSION TITLE
✨ Function parameter validation

### DISCUSSION THREAD

**Discussion by sagikazarmark** [dagger/dagger#12140] [Ideas]
### What are you trying to do?

Today, Dagger enforces "some" validation rules on function parameters. For example, there is a way to define enums and enforce that a parameter passed to a function is a valid enum.

However, there are way more cases where validating input values would make sense and better support built into the SDKs would make module code more concise and easier to read.

Some examples:

- version number
- enums that are hard to represent as enums right now (eg. a limited set of version numbers)
- special formatted strings

Some ideas for builtin validation support:

**Validation pragma**

```go
func myFunc(
    // +validate="[0-9]*\.[0-9]*\.[0-9]*"
    version string,
) {

}
```

**Custom type implementing an interface**

(example is Go specific, but I guess it's doable in other languages)

```go
type Version string

// SDK could call this function automatically
func (v Version) Validate() error {
    if v != "1.0.0" {
        return errors.New("invalid version")
    }

    return nil
}

func myFunc(
    version Version,
) {

}
```

### Why is this important to you?

Function code becomes easier to read:

```go
func myFunc(
    version string,
) { // no need to return an error
    // no need to manually validate version here
}
```

### How are you currently working around this?

I build validation code into modules, but that:

- makes functions longer
- requires functions to return an error type

**Comment by jedevc** [comment]
cc @shykes @helderco wdyt? I quite like this, feels like a pretty easy way to improve module readability.

**Comment by sagikazarmark** [comment]
Personally, I like the validation interface better. The pragma may be more universal across different SDKs on the surface, but the accepted values there may not be that easy unify.

For example, some Go validation libraries support struct tag based validation (which is similar) and uses a language like this: `notNull,mandatory,constraints:[StringNoControlCharacters{},StringLength{Minimum: 1, Maximum: 255}]`

Supporting only regex can quickly become uncomfortable, implementing a more sophisticated solution that is universal across SDKs may not be easy.

The interface based solution _should_ (tm) be fairly universal and is the more flexible one IMO.

**Comment by jasonmccallister** [comment]
Just going to add some comments here as I was going to post this feature request this morning as I was working on a module and needed validation.

```go
        // version must be a valid PHP version
	switch version {
	case "8.2", "8.3", "8.4":
	default:
		return nil, errors.New("invalid PHP version")
	}
```

Having syntax that supported validation inline would be immensely useful for module developers. 

I've used a few validation libraries and packages but the one syntax that I have come to appreciate the most is the [Laravel validation library](https://laravel.com/docs/11.x/validation#available-validation-rules), mostly because it can be expressed as string syntax and the framework will do the legwork behind the scenes. A similar approach might make sense given the documentation for arguments is in comment form.

Some examples from the Laravel documentation:

```
digits_between:1,100
```

```
ends_with:foo,bar...
```

```
in:8.2,8.3,8.4
```

They do support custom validation rules so an interface would be really great like @sagikazarmark is suggesting above.

Also, it might be useful to support multiple `// +validate=""` definitions (if possible) to avoid trying to join them on a special character (e.g. `// +validate="<one>;<two>"`. I feel like that syntax would get way too verbose and be difficult to parse.
